src/middleware/compression.rs

```diff
@@ -0,0 +1,239 @@
+#![cfg(feature = "compression")]
+use brotli2::read::{BrotliDecoder, BrotliEncoder};
```

Comment:
I think we should use `brotli2::bufread` instead of `brotli2::read`, IIRC buffer based read/write are more efficient, not sure if it applies here.

Reply:
Yeah, it should be -- I was just going off whatever encoders/decoders used `std::io::Read` as their interface (as Bytes exposes a `Read` handle through the `IntoBuf` trait. But this approach isn't likely to work as following another discussion [here](https://github.com/rustasync/tide/pull/194#discussion_r279182624).

For now I'm sort of back at the drawing board, as if I use the low-level in-memory chunk encoder/decoder that both [flate2](https://docs.rs/flate2/1.0.7/flate2/struct.Compress.html) and [brotli2](https://docs.rs/brotli2/0.3.2/brotli2/raw/struct.Compress.html) expose, I'll have to come up with some sort of strategy for decoding potentially large blocks and returning the stream of `Result<Bytes, Error>` piecemeal, unless the Stream generated by a non-compressed body already does this for me (then in which case I can read all the way to EOF, although this might be highly inefficient if I'm wrong).

Reply:
Looking at this, one thing I’ve been wondering is why `body` is a `Stream<Item = Bytes>` instead of an `AsyncRead`. If this is based on implementation details of `hyper` then maybe it would be better to change `http-service` to pass an `AsyncRead` (or more likely `AsyncBufRead`), then there can easily be an adaptor crate that implements just the compression adaptors on top of that which could be used here.